cql3/statements/create_aggregate_statement.cc

```diff
@@ -0,0 +1,86 @@
+ * (at your option) any later version.
+ *
+ * Scylla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "cql3/statements/create_aggregate_statement.hh"
+#include "cql3/functions/functions.hh"
+#include "cql3/functions/user_aggregate.hh"
+#include "prepared_statement.hh"
+#include "service/migration_manager.hh"
+#include "service/storage_proxy.hh"
+#include "database.hh"
+#include "cql3/query_processor.hh"
+
+namespace cql3 {
+
+namespace statements {
+
+void create_aggregate_statement::create(service::storage_proxy& proxy, functions::function* old) const {
+    if (old && !dynamic_cast<functions::user_aggregate*>(old)) {
+        throw exceptions::invalid_request_exception(format("Cannot replace '{}' which is not a user defined aggregate", *old));
+    }
+    data_type state_type = prepare_type(proxy, *_stype);
+
+    auto&& db = proxy.get_db().local();
+    std::vector<data_type> acc_types{state_type};
+    acc_types.insert(acc_types.end(), _arg_types.begin(), _arg_types.end());
+    auto state_func = dynamic_pointer_cast<functions::scalar_function>(functions::functions::find(functions::function_name{_name.keyspace, _sfunc}, acc_types));
+    auto final_func = dynamic_pointer_cast<functions::scalar_function>(functions::functions::find(functions::function_name{_name.keyspace, _ffunc}, {state_type}));
+ 
+    if (!state_func) {
+        throw exceptions::invalid_request_exception(format("State function not found: {}", _sfunc));
+    }
+    if (!final_func) {
+        throw exceptions::invalid_request_exception(format("Final function not found: {}", _ffunc));
+    }
+
+    auto dummy_ident = ::make_shared<column_identifier>("", true);
+    auto column_spec = make_lw_shared<column_specification>("", "", dummy_ident, state_type);
+    auto initcond_term = _ival->prepare(db, _name.keyspace, {column_spec});
+    bytes_opt initcond = to_bytes(*to_managed_bytes_opt(initcond_term->bind_and_get(cql3::query_options::DEFAULT)));
+
+    _aggregate = ::make_shared<functions::user_aggregate>(_name, initcond, std::move(state_func), std::move(final_func));
+    return;
+}
+
+std::unique_ptr<prepared_statement> create_aggregate_statement::prepare(database& db, cql_stats& stats) {
+    return std::make_unique<prepared_statement>(make_shared<create_aggregate_statement>(*this));
+}
+
+future<shared_ptr<cql_transport::event::schema_change>> create_aggregate_statement::announce_migration(
+        query_processor& qp) const {
+    if (!_aggregate) {
+        return make_ready_future<::shared_ptr<cql_transport::event::schema_change>>();
+    }
+    return qp.get_migration_manager().announce_new_aggregate(_aggregate).then([this] {
```

Comment:
Should we block this on a schema feature? It's painful, but on the other hand someone creating an aggregate before the cluster is fully upgraded can easily confuse it beyond all repair.

Reply:
Probably, we didn't create any feature for UDF-s though, and it's a nearly identical case. Perhaps we ignored the schema feature problem because the support is still marked as experimental? Or maybe not doing so was a bug.

Reply:
Oh wait, we didn't create a **schema** feature for UDF-s, but there is a cluster feature. So I'll just add one for UDA as well.